core/registration.c

```diff
@@ -191,30 +199,102 @@ static void prv_handleRegistrationReply(lwm2m_transaction_t * transacP,
                                         void * message)
 {
     coap_packet_t * packet = (coap_packet_t *)message;
-    lwm2m_server_t * targetP = (lwm2m_server_t *)(transacP->userData);
+    lwm2m_context_t * contextP = (lwm2m_context_t *)transacP->userData;
+    lwm2m_server_t * targetP = NULL;
+
+    if (contextP != NULL) targetP = utils_findServer(contextP, transacP->peerH);
 
-    if (targetP->status == STATE_REG_PENDING)
+    if (targetP != NULL && targetP->status == STATE_REG_PENDING)
     {
         time_t tv_sec = lwm2m_gettime();
         if (tv_sec >= 0)
         {
             targetP->registration = tv_sec;
         }
-        if (packet != NULL && packet->code == COAP_201_CREATED)
+        /* handle reply for block transfer of registration request */
+        if (packet != NULL && IS_OPTION(packet, COAP_OPTION_BLOCK1))
         {
-            targetP->status = STATE_REGISTERED;
-            if (NULL != targetP->location)
+            /* we should send next blocks only if we get success */
+            if ((packet->code & 0xE0) == 2 << 5)
             {
-                lwm2m_free(targetP->location);
+                uint32_t block1_num;
+                uint8_t  block1_more;
+                uint16_t block1_size;
+                uint32_t block1_offset;
+                coap_get_header_block1(packet, &block1_num, &block1_more, &block1_size, &block1_offset);
+                targetP->rem_payload_length -= block1_size;
```

Comment:
I think this should subtract targetP->block_size instead of block1_size. That represents how much was sent.